### PR TITLE
Add use of global flag 'use_mkldnn' to layer_helper

### PR DESCRIPTION
**Pull request by arlesniak**
### PR types

New features
### PR changes

Others
### Describe

Get use of global  FLAGS_use_mkldnn to append_activation in layer_helper

### CODE REVIEW

**Comment by arlesniak** [comment]
PR-CI-Coverage does not update status but after clicking Details it's all green

File: python/paddle/fluid/layer_helper.py [hunk 1]
```diff
@@ -147,8 +147,13 @@ def append_activation(self, input_var):
 
         if 'use_cudnn' in self.kwargs and self.kwargs.get('use_cudnn'):
             act['use_cudnn'] = self.kwargs.get('use_cudnn')
-        if 'use_mkldnn' in self.kwargs:
-            act['use_mkldnn'] = self.kwargs.get('use_mkldnn')
+
+        # priority order: use_mkldnn in kwargs then global
+        use_mkldnn = self.kwargs.get(
+            'use_mkldnn', core.globals().get("FLAGS_use_mkldnn", False))
+        if (use_mkldnn is not None) and use_mkldnn:
```
Comment (wojtuss) [inline python/paddle/fluid/layer_helper.py]: This is equivalent to `if use_mkldnn:`.

**Comment by luotao1** [comment]
> PR-CI-Coverage does not update status but after clicking Details it's all green

Some network error. If you will fix https://github.com/PaddlePaddle/Paddle/pull/26497#discussion_r478162539, you can push a new commit to trigger it.

**Review by grygielski — DISMISSED** [review]
LGTM

**Comment by luotao1** [comment]
```
2020-08-27 19:32:07 ****************
2020-08-27 19:32:07 0. You must have one RD (zhiqiu (Recommend) or phlrain) approval for the api change for the opreator-related api without 'core.ops'.
2020-08-27 19:32:07 For more details, please click [https://github.com/PaddlePaddle/Paddle/wiki/paddle_api_development_manual.md]
2020-08-27 19:32:07 Related APIs: paddle.common_ops_import.LayerHelper.append_activation
2020-08-27 19:32:07 paddle.common_ops_import.LayerHelper.append_op
2020-08-27 19:32:07 paddle.fluid.layers.utils.LayerHelper.append_activation
2020-08-27 19:32:07 paddle.fluid.layers.utils.LayerHelper.append_op
2020-08-27 19:32:07 
2020-08-27 19:32:07 There are 1 approved errors.
2020-08-27 19:32:07 ****************
```

File: python/paddle/fluid/layer_helper.py (hunk 1 above)
Comment (arlesniak) [reply python/paddle/fluid/layer_helper.py]: Done

File: python/paddle/fluid/layer_helper.py [hunk 2]
```diff
@@ -40,7 +42,10 @@ def __init__(self, layer_type, **kwargs):
             self.kwargs['name'], layer_type=layer_type)
 
     def append_op(self, *args, **kwargs):
-        return self.main_program.current_block().append_op(*args, **kwargs)
+        op = self.main_program.current_block().append_op(*args, **kwargs)
+        if self.debug_:
+            self.appended_op = op
+        return op
```
Comment (zhiqiu) [inline python/paddle/fluid/layer_helper.py]: It seems `appended_op` is only used to get the operator for testing its attrs. 
I think we don't need to add a member to LayerHelper, the first appended op (in your test case, only one op is appended) can be obtained by `fluid.default_main_program().global_block().ops[0] `.

File: python/paddle/fluid/tests/unittests/test_imperative_basic.py [hunk 3]
```diff
@@ -640,6 +641,37 @@ def test_append_activation_in_dygraph3(self):
...
+        with fluid.dygraph.guard():
+            a = fluid.dygraph.to_variable(a_np)
+            res1 = func(a)
+            res2 = fluid.layers.relu(a)
+            self.assertTrue(helper.appended_op.attrs.get('use_mkldnn', False))
+            self.assertTrue(np.array_equal(res1.numpy(), res2.numpy()))
+
+    def test_append_activation_in_dygraph_global_use_mkldnn(self):
+        a_np = np.random.uniform(-2, 2, (10, 20, 30)).astype(np.float32)
+        fluid.set_flags({'FLAGS_use_mkldnn': True})
```
Comment (wojtuss) [inline python/paddle/fluid/tests/unittests/test_imperative_basic.py]: @luotao1 Here I have the same question as [here](https://github.com/PaddlePaddle/Paddle/pull/25774#discussion_r462963970), whether setting the global flag can interfere with other tests running in parallel on CI.
Other than that, LGTM.

File: python/paddle/fluid/layer_helper.py [hunk 4]
```diff
@@ -147,8 +152,13 @@ def append_activation(self, input_var):
...
             act['use_cudnn'] = self.kwargs.get('use_cudnn')
-        if 'use_mkldnn' in self.kwargs:
-            act['use_mkldnn'] = self.kwargs.get('use_mkldnn')
+
+        # priority order: use_mkldnn in kwargs then global
+        use_mkldnn = self.kwargs.get(
+            'use_mkldnn', core.globals().get("FLAGS_use_mkldnn", False))
+        if use_mkldnn:
+            act['use_mkldnn'] = use_mkldnn
+
```
Comment (zhiqiu) [inline python/paddle/fluid/layer_helper.py]: Maybe it is better to implement this functionality in C++ side. There are 2 reasons,

1. `FLAGS_use_mkldnn ` is a GFLAG, which is declared and used in c++. 
2. There may be other way to add an activation op, for example, `_append_activation_in_dygraph`.

File: python/paddle/fluid/layer_helper.py (hunk 2 above)
Comment (arlesniak) [reply python/paddle/fluid/layer_helper.py]: Thank you for the comment 
> fluid.default_main_program().global_block().ops

is empty in dygraph mode, I'm working on different solution.

**Review by grygielski — APPROVED** [review]
LGTM

File: paddle/fluid/imperative/tracer.cc [hunk 5]
```diff
@@ -47,6 +49,9 @@ void Tracer::TraceOp(const std::string& type, const NameVarBaseMap& ins,
                      const NameVarBaseMap& outs, framework::AttributeMap attrs,
                      const platform::Place& place, bool trace_backward) {
   VLOG(1) << "Trace Op: " << type;
+  if (FLAGS_use_mkldnn) {
+    attrs["use_mkldnn"] = true;
+  }
```
Comment (zhiqiu) [inline paddle/fluid/imperative/tracer.cc:54]: LGTM for dygraph mode.
I don't see handling `use_mkldnn` in static mode, is it already done in other PR?

**Review by wojtuss — APPROVED** [review]
LGTM

**Comment by luotao1** [comment]
@zhiqiu Please have a double review!

File: python/paddle/fluid/tests/unittests/test_imperative_basic.py [hunk 6]
```diff
@@ -640,6 +641,31 @@ def test_append_activation_in_dygraph3(self):
...
+        helper = LayerHelper(
+            fluid.unique_name.generate("test"), act="relu", use_mkldnn=True)
+        func = helper.append_activation
+        with fluid.dygraph.guard():
+            a = fluid.dygraph.to_variable(a_np)
+            res1 = func(a)
+            res2 = fluid.layers.relu(a)
+            self.assertTrue(np.array_equal(res1.numpy(), res2.numpy()))
+
+    def test_append_activation_in_dygraph_global_use_mkldnn(self):
```
Comment (zhiqiu) [inline python/paddle/fluid/tests/unittests/test_imperative_basic.py:655]: LGTM for the tests to check the accuracy of the result when `use_mkldnn` is True and False.

I wonder if there are another ways that we checked the `mkldnn` is really used. For example, we can add some VLOG in the mkldnn kernel, and see if the log is printed when `fluid.set_flags({'FLAGS_use_mkldnn': True})` is called. And we can do it manually it is difficult for unittest.
Maybe you have already done that, I just wonder that.

File: python/paddle/fluid/tests/unittests/test_imperative_basic.py (hunk 3 above)
Comment (wojtuss) [reply python/paddle/fluid/tests/unittests/test_imperative_basic.py]: I can't see any option to test global flags in a different way, also other flags are tested this way. After the last commit the time spent with the flag set is minimum.

File: python/paddle/fluid/layer_helper.py (hunk 4 above)
Comment (arlesniak) [reply python/paddle/fluid/layer_helper.py]: I've changed tracer on the c++ side, so there is no need for the above changes anymore.

File: paddle/fluid/imperative/tracer.cc (hunk 5 above)
Comment (arlesniak) [reply paddle/fluid/imperative/tracer.cc:54]: If you mean running through executor, it's already done in executor.cc:185

File: python/paddle/fluid/tests/unittests/test_imperative_basic.py (hunk 6 above)
Comment (arlesniak) [reply python/paddle/fluid/tests/unittests/test_imperative_basic.py:655]: The problem we had is that AFAIK there is no better way to check if mkldnn kernel has been run than to check the output of DNNL_VERBOSE for now. The other idea is to run kernel throwing Exception with extended information about type of kernel. The other problem is that AFAIK python unittest lib does not have parsing c++ stdout result out of the box. Do you have suggestion what is the best way to check what type of kernel has been run from the python level ?

Comment (luotao1) [reply python/paddle/fluid/tests/unittests/test_imperative_basic.py:655]: > Do you have suggestion what is the best way to check what type of kernel has been run from the python level

I think checking the output of DNNL_VERBOSE is Ok. You can do it in the next PR.

**Pull request merged**